api/src/sparql.rs

```diff
@@ -0,0 +1,104 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//!
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::dataset::Dataset;
+use crate::graph::Graph;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset: Dataset {
+    type BindingsTerm: TTerm;
+    type BindingsResult: SparqlBindings<Self::BindingsTerm>;
+    type GraphResult: Graph;
+    type SparqlError: Error + 'static;
+
+    /// Parse and immediately execute `query`
+    fn query(&self, query: &str) -> Result<SparqlResult<Self>, Self::SparqlError>;
+}
+
+/// The result of executing a SPARQL query.
+pub enum SparqlResult<T>
+where
+    T: SparqlDataset + ?Sized,
+{
+    Bindings(T::BindingsResult),
+    Boolean(bool),
+    Graph(T::GraphResult),
+}
+
+/// The result of executing a SPARQL SELECT query
+pub trait SparqlBindings<T: TTerm>: IntoIterator<Item = Vec<Option<T>>> {
+    /// Borrow the list of SELECTed variable names
+    fn variables(&self) -> &[&str];
```

Comment:
As you stated in the other PR, this requires the bindings to own the `[&str]` slice. This might not be the case of all implementations that might own a `Vec<String>` or a `Vec<Variable>`.

Reply:
you are right, I got carried away ;)